File: frontend/src/api/memory.ts
import { apiGet, apiPost } from "./http";

export function createMemory(payload: Record<string, unknown>) {
  return apiPost<Record<string, unknown>>("/api/v1/memory/notes", payload);
}

export function searchMemory(payload: Record<string, unknown>) {
  return apiPost<Record<string, unknown>>("/api/v1/memory/search", payload);
}

export function getMemoryNote(noteId: number) {
  return apiGet<Record<string, unknown>>(`/api/v1/memory/notes/${noteId}`);
}

export function createMemoryLink(payload: Record<string, unknown>) {
  return apiPost<Record<string, unknown>>("/api/v1/memory/links", payload);
}

export function buildReflectionFromReview(reviewId: number) {
  return apiPost<Record<string, unknown>>(`/api/v1/memory/reflection/from-review/${reviewId}`);
}

export function buildPlaybookFromReview(reviewId: number) {
  return apiPost<Record<string, unknown>>(`/api/v1/memory/playbook/from-review/${reviewId}`);
}

export function listPlaybooks() {
  return apiGet<Record<string, unknown>>("/api/v1/memory/playbooks");
}

export function disableMemory(noteId: number, reason?: string) {
  return apiPost<Record<string, unknown>>(`/api/v1/memory/notes/${noteId}/disable`, { reason });
}

export function markConflict(noteId: number, status: string, reason?: string) {
  return apiPost<Record<string, unknown>>(`/api/v1/memory/notes/${noteId}/conflict`, { status, reason });
}

export function getMemoryConfig() {
  return apiGet<Record<string, unknown>>("/api/v1/memory/config");
}

File: backend/api/v1/memory.py
from __future__ import annotations

from fastapi import APIRouter, Request

from backend.core.config import get_app_config
from backend.core.responses import success_response
from backend.core.security import sanitize_config
from memory.schemas import (
    ConflictMemoryRequest,
    DisableMemoryRequest,
    MemoryLinkCreate,
    MemoryNoteCreate,
    MemorySearchQuery,
)
from memory.service import MemoryService


router = APIRouter(prefix="/memory", tags=["memory"])


@router.post("/notes")
def create_memory_note(payload: MemoryNoteCreate, request: Request) -> dict:
    data = MemoryService().create_memory(payload).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.get("/notes/{note_id}")
def get_memory_note(note_id: int, request: Request) -> dict:
    data = MemoryService().get_memory(note_id).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/search")
def search_memory(payload: MemorySearchQuery, request: Request) -> dict:
    result = MemoryService().search_memory(payload)
    data = result.model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/links")
def create_memory_link(payload: MemoryLinkCreate, request: Request) -> dict:
    data = MemoryService().create_link(payload).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/reflection/from-review/{review_id}")
def create_reflection_from_review(review_id: int, request: Request) -> dict:
    data = MemoryService().build_reflection_from_review(review_id).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/playbook/from-review/{review_id}")
def create_playbook_from_review(review_id: int, request: Request) -> dict:
    data = MemoryService().build_playbook_from_review(review_id).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.get("/playbooks")
def list_playbooks(request: Request) -> dict:
    playbooks = MemoryService().list_playbooks()
    data = {
        "playbooks": [playbook.model_dump(mode="json") for playbook in playbooks],
        "count": len(playbooks),
        "real_trading_enabled": get_app_config().real_trading_enabled,
    }
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/notes/{note_id}/disable")
def disable_memory_note(note_id: int, payload: DisableMemoryRequest, request: Request) -> dict:
    data = MemoryService().disable_memory(note_id, reason=payload.reason).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.post("/notes/{note_id}/conflict")
def mark_memory_conflict(note_id: int, payload: ConflictMemoryRequest, request: Request) -> dict:
    data = MemoryService().mark_memory_conflict(note_id, status=payload.status, reason=payload.reason).model_dump(mode="json")
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)


@router.get("/config")
def memory_config(request: Request) -> dict:
    data = MemoryService().config_summary()
    data["real_trading_enabled"] = get_app_config().real_trading_enabled
    return success_response(data=sanitize_config(data), trace_id=request.state.trace_id)
